feat: add `number/float16/base/div`

Lokeshranjan8 opened the pull request:
Resolves none.

## Description

> What is the purpose of this pull request?

This pull request: 

-   Adds number/float16/base/div

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

### AI Assistance

> When authoring the changes proposed in this PR, did you use any kind of AI assistance?

-   [ ] Yes
-   [x] No


* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Lokeshranjan8 committed:
feat: add number/float16/base/div

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

Lokeshranjan8 committed:
Update copyright year from 2023 to 2025

Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Lokeshranjan8 committed:
Clean up package.json by removing gypfile and directories

Removed unnecessary properties from package.json.

Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Neerajpathak07 commented on lib/node_modules/@stdlib/number/float16/base/div/lib/main.js:
```diff
@@ -0,0 +1,62 @@
+// MAIN //
+
+/**
+* Divides two half-precision floating-point numbers `x` and `y`.
```
```suggestion
* Divides two half-precision floating-point numbers.
```

Neerajpathak07 replied:
Description should be consistent throughout the JS, TS & repl files!!

Neerajpathak07 commented on lib/node_modules/@stdlib/number/float16/base/div/README.md:
```diff
@@ -0,0 +1,113 @@
+
+#### div( x, y )
+
+divides two half-precision floating-point numbers.
```
```suggestion
Divides two half-precision floating-point numbers.
```

Neerajpathak07 commented on lib/node_modules/@stdlib/number/float16/base/div/examples/index.js:
```diff
@@ -0,0 +1,28 @@
+var x = discreteUniform( 100, -50, 50 );
+var y = discreteUniform( 100, -50, 50 );
+
+logEachMap( '%d / %d = %d', x, y, div );
```
```javascript
var toFloat16 = require( '@stdlib/number/float64/base/to-float16' );
var uniform = require( '@stdlib/random/array/uniform' );
var map = require( '@stdlib/array/base/map' );
var logEachMap = require( '@stdlib/console/log-each-map' );
var div= require( './../lib' );

// Create an array of random half-precision floating-point numbers:
var x = map( uniform( 100, -50.0, 50.0 ), toFloat16 );

// Create an array of random half-precision floating-point numbers:
var y = map( uniform( 100, -50.0, 50.0 ), toFloat16 );

logEachMap( 'x: %0.4f, y: %0.4f => div: %0.4f', x, y, div);
```
Wondering if using `map` would be better than `discreteUniform` as by using `map` we can directly store half-precision numbers. Hence, providing more clarity
cc:- @kgryte

kgryte replied:
Yes, this is better.

Lokeshranjan8 committed:
Update lib/node_modules/@stdlib/number/float16/base/div/lib/main.js

Co-authored-by: Neeraj Pathak <neerajrpathak710@gmail.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Lokeshranjan8 committed:
Update lib/node_modules/@stdlib/number/float16/base/div/README.md

Co-authored-by: Neeraj Pathak <neerajrpathak710@gmail.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Lokeshranjan8 committed:
bench: generate half-precision inputs using uniform distribution

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

Lokeshranjan8 committed:
bench: wrap toFloat16 with naryFunction in map

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Lokeshranjan8 committed:
Revise div function documentation for clarity

Updated documentation to remove reference to 'y'.

Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Neerajpathak07 commented on lib/node_modules/@stdlib/number/float16/base/div/package.json:
```diff
@@ -0,0 +1,180 @@
+{
+  "name": "@stdlib/number/float16/base/div",
+  "version": "0.0.0",
+  "description": "divide two half-precision floating-point numbers.",
```
```suggestion
  "description": "Divide two half-precision floating-point numbers.",
```

Neerajpathak07 commented on lib/node_modules/@stdlib/number/float16/base/div/package.json:
```diff
@@ -0,0 +1,180 @@
+      "alias": "div",
+      "pkg_desc": "division of two half-precision floating-point numbers",
+      "desc": "division of two half-precision floating-point numbers",
+      "short_desc": "division",
```
```suggestion
      "short_desc": "",
```

Neerajpathak07 commented:
All-in-all looks good just the comment on `package.json` descriptions.

Lokeshranjan8 committed:
Update lib/node_modules/@stdlib/number/float16/base/div/package.json

Co-authored-by: Neeraj Pathak <neerajrpathak710@gmail.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Lokeshranjan8 committed:
Update lib/node_modules/@stdlib/number/float16/base/div/package.json

Co-authored-by: Neeraj Pathak <neerajrpathak710@gmail.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Planeshifter commented on lib/node_modules/@stdlib/number/float16/base/div/examples/index.js:
```diff
@@ -0,0 +1,30 @@
+var uniform = require( '@stdlib/random/array/uniform' );
+var map = require( '@stdlib/array/base/map' );
+var logEachMap = require( '@stdlib/console/log-each-map' );
+var div= require( './../lib' );
```
Missing space before `=` in variable assignment. Per the stdlib style guide, there should be spacing around operators.

```suggestion
var div = require( './../lib' );
```

Planeshifter commented on lib/node_modules/@stdlib/number/float16/base/div/examples/index.js:
```diff
@@ -0,0 +1,30 @@
+var x = map( uniform( 100, -50.0, 50.0 ), toFloat16 );
+var y = map( uniform( 100, -50.0, 50.0 ), toFloat16 );
+
+logEachMap( 'x: %0.4f, y: %0.4f => div: %0.4f', x, y, div);
```
Missing space before closing parenthesis. Per the stdlib style guide, function calls should follow the pattern `foo( bar )` with spaces inside the parentheses.

```suggestion
logEachMap( 'x: %0.4f, y: %0.4f => div: %0.4f', x, y, div );
```

Planeshifter commented on lib/node_modules/@stdlib/number/float16/base/div/README.md:
```diff
@@ -0,0 +1,115 @@
+var x = map( uniform( 100, -50.0, 50.0 ), toFloat16 );
+var y = map( uniform( 100, -50.0, 50.0 ), toFloat16 );
+
+logEachMap( 'x: %0.4f, y: %0.4f => div: %0.4f', x, y, div);
```
Missing space before closing parenthesis. Per the stdlib style guide, function calls should follow the pattern `foo( bar )` with spaces inside the parentheses.

```suggestion
logEachMap( 'x: %0.4f, y: %0.4f => div: %0.4f', x, y, div );
```

Planeshifter commented on lib/node_modules/@stdlib/number/float16/base/div/test/test.js:
```diff
@@ -0,0 +1,102 @@
+var tape = require( 'tape' );
+var isNegativeZero = require( '@stdlib/math/base/assert/is-negative-zerof' );
+var isPositiveZero = require( '@stdlib/math/base/assert/is-positive-zerof' );
+var isnan = require( '@stdlib/math/base/assert/is-nanf' );
```
@kgryte In https://github.com/stdlib-js/stdlib/commit/79f62f06d15dfbadca560345f9857d7367717c25, the float64 functions are used in the tests:

```js
var isNegativeZero = require( '@stdlib/math/base/assert/is-negative-zero' );
var isPositiveZero = require( '@stdlib/math/base/assert/is-positive-zero' );
var isnan = require( '@stdlib/math/base/assert/is-nan' );
```

I assume we should do the same here given we do not have any such utilities for half-precision numbers?

Neerajpathak07 replied:
ref:- https://github.com/stdlib-js/stdlib/pull/9442#discussion_r2663368574

kgryte replied:
@Planeshifter These are in the process of being created.

Planeshifter commented on lib/node_modules/@stdlib/number/float16/base/div/test/test.js:
```diff
@@ -0,0 +1,102 @@
+var isPositiveZero = require( '@stdlib/math/base/assert/is-positive-zerof' );
+var isnan = require( '@stdlib/math/base/assert/is-nanf' );
+var PINF = require( '@stdlib/constants/float32/pinf' );
+var NINF = require( '@stdlib/constants/float32/ninf' );
```
Use `float16` constants instead of `float32` constants.

```suggestion
var PINF = require( '@stdlib/constants/float16/pinf' );
var NINF = require( '@stdlib/constants/float16/ninf' );
```

Planeshifter reviewed (commented):
Left some more comments. Should be close to landing!

Lokeshranjan8 committed:
Update lib/node_modules/@stdlib/number/float16/base/div/examples/index.js

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Lokeshranjan8 committed:
Update lib/node_modules/@stdlib/number/float16/base/div/examples/index.js

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Lokeshranjan8 committed:
Update lib/node_modules/@stdlib/number/float16/base/div/README.md

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Lokeshranjan8 committed:
Update lib/node_modules/@stdlib/number/float16/base/div/test/test.js

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Planeshifter commented on lib/node_modules/@stdlib/number/float16/base/div/test/test.js:
```diff
@@ -0,0 +1,102 @@
+var tape = require( 'tape' );
+var isNegativeZero = require( '@stdlib/math/base/assert/is-negative-zerof' );
+var isPositiveZero = require( '@stdlib/math/base/assert/is-positive-zerof' );
+var isnan = require( '@stdlib/math/base/assert/is-nanf' );
```
These are the single-precision (`*f`) variants of the assertion functions, but the reference package (`float16/base/mul`) uses the double-precision versions (`is-negative-zero`, `is-positive-zero`, `is-nan`). Since `float64ToFloat16` returns a JavaScript number (which is float64), we should use the same assertion functions as `mul`:

```suggestion
var isNegativeZero = require( '@stdlib/math/base/assert/is-negative-zero' );
var isPositiveZero = require( '@stdlib/math/base/assert/is-positive-zero' );
var isnan = require( '@stdlib/math/base/assert/is-nan' );
```

Neerajpathak07 replied:
Let's update these variable paths and wire them from `number/float16/base/assert`

Planeshifter commented on lib/node_modules/@stdlib/number/float16/base/div/docs/types/index.d.ts:
```diff
@@ -0,0 +1,53 @@
+* var v = div( -1.0, 5.0 );
+* // returns ~-0.2
+*
+* @example
+* v = div( 2.0, 5.0 );
+* // returns ~0.4
+*
+* @example
+* v = div( 0.0, 5.0 );
+* // returns 0.0
+*
+* @example
+* v = div( -0.0, 5.0 );
+* // returns -0.0
+*
+* @example
+* v = div( NaN, NaN );
+* // returns NaN
```
These examples are missing the `var` keyword, since each example has its own scope:

```suggestion
* @example
* var v = div( 2.0, 5.0 );
* // returns ~0.4
*
* @example
* var v = div( 0.0, 5.0 );
* // returns 0.0
*
* @example
* var v = div( -0.0, 5.0 );
* // returns -0.0
*
* @example
* var v = div( NaN, NaN );
* // returns NaN
```

Lokeshranjan8 committed:
Update lib/node_modules/@stdlib/number/float16/base/div/docs/types/index.d.ts

Co-authored-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Neerajpathak07 commented on lib/node_modules/@stdlib/number/float16/base/div/package.json:
```diff
@@ -0,0 +1,180 @@
+      "base_alias": "div",
+      "alias": "div",
+      "pkg_desc": "division of two half-precision floating-point numbers",
+      "desc": "division of two half-precision floating-point numbers",
```
```suggestion
      "pkg_desc": "divide two half-precision floating-point numbers",
      "desc": "divides two half-precision floating-point numbers",
```

Neerajpathak07 reviewed (changes requested):
@Lokeshranjan8 All-in-All this PR is in good shape just a few comments to address!!

Lokeshranjan8 committed:
Update lib/node_modules/@stdlib/number/float16/base/div/package.json

Co-authored-by: Neeraj Pathak <neerajrpathak710@gmail.com>
Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Lokeshranjan8 committed:
Update test.js

Signed-off-by: Lokesh Ranjan <lokesh.ranjan.phe23@itbhu.ac.in>

Lokeshranjan8 committed:
Merge branch 'develop' into for-loop-utils
